### PR TITLE
More details on the RPC part of replication

### PR DESCRIPTION
**Pull request by staltz**
This PR is distilling the ideas @arj03 and I got in a video call yesterday, in proper text form. A rough draft of these ideas is [over there](https://github.com/ssb-ngi-pointer/ssb-replication-scheduler/pull/5#issuecomment-900376809).

### CODE REVIEW

File: README.md
Comment (staltz) [inline README.md]: Problem I just discovered: what if you already have gotten the meta feed ID of your friend in the past, but then restarted sbot, are you going to use `getSubset` *again* to re-discover it? It would be smarter that your sbot remembers the association between friend's main feed IDs and their root meta feed IDs.

Further, it doesn't seem sustainable to rely on `getSubset` for this, because then it means that a peer that wants to do partial replication *must* be connected to another peer that is doing full replication, because they will possess the `metafeed/announce` messages.

Thinking out loud: I know! We just need to `addOOO` the `metafeed/announce` message locally. That's perhaps important to amend in this PR.

**Comment by arj03** [comment]
Some comments while I read along on this. 

I think we should remove fusionidentities from the dictionary as that will make it easier to see exactly what this is about

**Comment by arj03** [comment]
On indexes:

> They should only be used when synchonizing a feed for the first time. After this, EBT should be used to keep the feed in sync.

Need to revisit this statement

Same with:

> These index feeds can be replicated ....

**Comment by arj03** [comment]
> That is, we need new events in EBT that merely signal what is the clock state of feeds, without immediately requesting updates for those feeds. 

I would say this is where a `getClock` method on EBT comes into play. That should allow us to simply send a vector of *fake* starting points.

**Comment by arj03** [comment]
> To handle this, we first replicate index feeds in full, and persist the messages to the log. Then, we run a variant of EBT where getAt is implemented in terms of the index feed messages, and append will do a database addOOO for messages on the target feed. This EBT variant will require its own replicate duplex RPC in a separate muxrpc namespace not shared with the conventional ebt.replicate, because the data passed through the duplex will be different.

I'm starting to warm up to the @cryptix idea of being able to register formats to EBT. With this we could register a index format and have it send the index messages + what is indexed as pairs. Needed for bendy butt binary as well.

**Comment by arj03** [comment]
> If connected peers do not support this RPC, then we can fallback to createHistoryStream to fetch the friend's full feed in memory (i.e. without persisting it to the log) and then filter for the first message of type 'metafeed/announce'. Once we find that message, we persist it in the log (in out-of-order fashion), and we know the feed ID for our friend's root meta feed, enabling us to replicate it in full, like we do with any other feed.

It sounds like we can just start with this and continue once we have that 1 message. But we really can't continue because the other end will probably also not have the EBT changes we need. So I think we need to split this into partial replication given an old (patchwork) client and then partial replication given a newer client.

For the partial replication given an old client we already have all the messages we need for the main feed, so we could use that to store the slices of messages we need. One problem with that is that we don't have any kind of transaction support at the database layer so we could potentially be in a situation where this fails in the middle. And when we restart we could be connected to either a new client or an old client increasing the number of scenarios we need to take into account. The beauty of existing replication (and index replication for that sake) is that we can trust that things only ever increase. Come to think of it, the transaction argument is true in the index replication case as well. I wonder if that means it is still better to do index replication as 2 streams like @cryptix mentioned.

**Comment by arj03** [comment]
I think we need a section 5 on frontier replication related to EBT.

**Comment by arj03** [comment]
Thinking more about this partial replication given an old client I think we could go down the store the slices of messages. A few things to note: we still try to do meta feed replication including indexes once we are connected to a newer client. We should be able to detect this kind of partial replication in the case of a crash and even if we are connected to a newer client should finish this replication. Then we should be able to continue on with step 2 wonce connected to a newer client. Note the index feed replication would be a simple replication of the feed not the index + indexed.

**Comment by arj03** [comment]
I think that was it for comments and brain dump of thinking process

**Review by arj03 — APPROVED** [review]
Great, I like this new structure

**Comment by staltz** [comment]
Do we wait for @cryptix's review before merging?

**Comment by arj03** [comment]
I think it would be good to have his blessing. I roughly went over the changes friday with him so I don't think there will be any major objections. I hope to start up on ebt changes later this week, so not much rush other than closing open PRs :)

**Comment by arj03** [comment]
I was looking over my notes from Svendsolar and can see that the steps outlined here are general enough that they should fit that model as well. Seeing that we had a session where this model was explained and there was no objections I think we are good to go merging this.

**Pull request merged**
